src/code_08026044.c

```diff
@@ -0,0 +1,443 @@
+            sub_0815604C(sprite1);
+            array2[0] = 0;
+            array2[1] = val;
+            array2[2] = val;
+            array2[3] = array[0];
+            array2[4] = array[1]; // UB: out-of-bounds access
+            arg0->unkCC.unk8 |= 0x2020;
+            sub_08155604(sprite, array2);
+            sub_0815604C(sprite);
+        }
+    }
+    if (arg0->flags & 0x8000) {
+        s16 array2[5];
+        s16 array[2];
+        sprite = &arg0->unkF4;
+        if (arg0->flags >= 0) {
+            if (!(arg0->flags & 0x10000) && !sub_08155128(sprite)) {
+                arg0->flags |= 0x10000;
+            }
+            if (arg0->flags & 0x10000 && arg0->flags & 0x20000) {
+                arg0->flags &= ~0x10000;
+                sprite->unk1B = 0xff;
+            }
+        }
+        if (arg0->flags & 0x40000) {
+            Macro_08026168(sprite, arg0->unk344, arg0->unk33C, arg0->unk340, array, array2, 0x42022);
+        }
+        sub_0815604C(sprite);
+    }
+    if (arg0->flags & 0x80000) {
+        sprite = &arg0->unk11C;
+        if (arg0->flags >= 0) {
+            if (!(arg0->flags & 0x100000) && !sub_08155128(sprite)) {
+                arg0->flags |= 0x100000;
+            }
+            if (arg0->flags & 0x100000 && arg0->flags & 0x200000) {
+                arg0->flags &= ~0x100000;
+                sprite->unk1B = 0xff;
+            }
+        }
+        sub_0815604C(sprite);
+    }
+}
+
+void sub_08026610(struct Task *task) {
+    struct Cutscene *cs2 = TaskGetStructPtr(task), *cs = cs2;
+    u16 i;
+    for (i = 0; i < 4; ++i) {
+        VramFree(cs->unk4[i].tilesVram);
+    }
+    VramFree(cs->unkA4.tilesVram);
+    VramFree(cs->unkCC.tilesVram);
+    VramFree(cs->unkF4.tilesVram);
+    VramFree(cs->unk11C.tilesVram);
+    VramFree(cs->unk144.tilesVram);
+    VramFree(cs->unk16C.tilesVram);
+}
+
+extern const struct Object11_8 *gUnk_082EB2B4;
+extern u16 gUnk_03000008;
+void sub_08029618(struct Cutscene *arg0);
+void sub_08026698(void) {
+    struct Cutscene *cs2 = TaskGetStructPtr(gCurTask), *cs = cs2;
+    u16 a;
+    const u16 *palette;
+    if (gPressedKeys & 0xb)
+        cs->flags |= 0x40000000;
+    if (!(cs->flags & 0x40000000) || (cs->flags & 0x20000000)) {
+#ifndef NONMATCHING
+        asm("":::"r6");
+#endif
+        cs->unk0(cs);
+        if (gUnk_03000008) {
+            return;
+        }
+        if (cs->flags & 0x1000000) {
+            u8 paletteOffset = gUnk_082EB2B4->unk0 >> 8;
+            u16 paletteSize = gUnk_082EB2B4->unk0 >> 0x10;
+            const struct Object11_8_4 *const *tmp3 = gUnk_082EB2B4->unk4;
+
+#ifndef NONMATCHING
+            asm("":::"r8");
+#endif
+            if (cs->unk35E >= (*(u8 *)&gUnk_082EB2B4->unk0)) {
+                cs->unk35E = 0;
+                cs->unk360 = 0;
+            }
+            else {
+                a = tmp3[cs->unk35E]->unk0;
+                do palette = tmp3[cs->unk35E]->palette;
+                while (0);
```

Comment:
TBH I don't think putting while (0); on a new line is more readable. In this way while is visually separated from do, which can make it look like two independent statements